components/p3a/message_manager_unittest.cc

```diff
@@ -0,0 +1,330 @@
+  base::test::ScopedFeatureList scoped_feature_list_;
+  network::TestURLLoaderFactory url_loader_factory;
+  scoped_refptr<network::SharedURLLoaderFactory> shared_url_loader_factory;
+  P3AConfig p3a_config;
+  std::unique_ptr<MessageManager> message_manager;
+  TestingPrefServiceSimple local_state;
+
+  std::map<std::string, size_t> p3a_json_sent_metrics;
+  std::map<std::string, size_t> p2a_json_sent_metrics;
+
+  std::set<std::string> p3a_star_sent_messages;
+
+  bool info_request_made = false;
+  size_t points_requests_made = 0;
+
+  uint8_t current_epoch = kInitialEpoch;
+  base::Time next_epoch_time;
+
+ private:
+  base::StringPiece ExtractBodyFromRequest(
+      const network::ResourceRequest& request) {
+    return request.request_body->elements()
+        ->at(0)
+        .As<network::DataElementBytes>()
+        .AsStringPiece();
+  }
+
+  void StoreJsonMetricInMap(const network::ResourceRequest& request,
+                            bool is_p2a) {
+    base::StringPiece body = ExtractBodyFromRequest(request);
+    base::Value parsed_log = *base::JSONReader::Read(body);
+    std::string metric_name = *parsed_log.FindStringKey("metric_name");
+    int metric_value = *parsed_log.FindIntKey("metric_value");
+
+    if (is_p2a) {
+      EXPECT_EQ(p2a_json_sent_metrics.find(metric_name),
+                p2a_json_sent_metrics.end());
+      p2a_json_sent_metrics[metric_name] = metric_value;
+    } else {
+      EXPECT_EQ(p3a_json_sent_metrics.find(metric_name),
+                p3a_json_sent_metrics.end());
+      p3a_json_sent_metrics[metric_name] = metric_value;
+    }
+  }
+};
+
+TEST_F(P3AMessageManagerTest, UpdateLogsAndSendJson) {
+  std::vector<std::string> test_histograms = GetTestHistogramNames(3, 4);
+
+  for (size_t i = 0; i < test_histograms.size(); i++) {
+    message_manager->UpdateMetricValue(test_histograms[i], i + 1);
+  }
+
+  task_environment_.FastForwardBy(base::Seconds(kUploadIntervalSeconds * 50));
+
+  EXPECT_EQ(p3a_json_sent_metrics.size(), 3U);
+  EXPECT_EQ(p2a_json_sent_metrics.size(), 4U);
+
+  for (size_t i = 0; i < test_histograms.size(); i++) {
+    if (test_histograms[i].rfind(kP2APrefix, 0) == 0) {
+      ASSERT_EQ(p2a_json_sent_metrics.find(test_histograms[i])->second, i + 1);
+    } else {
+      ASSERT_EQ(p3a_json_sent_metrics.find(test_histograms[i])->second, i + 1);
+    }
+  }
+
+  ResetInterceptorStores();
+  task_environment_.FastForwardBy(base::Days(7) +
+                                  base::Seconds(kUploadIntervalSeconds * 50));
+
+  EXPECT_EQ(p3a_json_sent_metrics.size(), 3U);
+  EXPECT_EQ(p2a_json_sent_metrics.size(), 4U);
+  for (size_t i = 0; i < test_histograms.size(); i++) {
+    if (test_histograms[i].rfind(kP2APrefix, 0) == 0) {
+      ASSERT_EQ(p2a_json_sent_metrics.find(test_histograms[i])->second, i + 1);
+    } else {
+      ASSERT_EQ(p3a_json_sent_metrics.find(test_histograms[i])->second, i + 1);
+    }
+  }
+}
+
+TEST_F(P3AMessageManagerTest, UpdateLogsAndSendStar) {
```

Comment:
What happens if the server is not available or has an error? We should have a test for that